Step1 제출합니다.

hjlee83 opened the pull request.

hojinDev committed:
docs : readme.md

기능 요구사항과 관련된 내용을 README.md에 작성.

hojinDev committed:
feat(lotto): 로또 구매 기능 추가

로또샵에서 금액을 입력받아 로또를 구매할 수 있는 기능을 추가. 1000원 단위가 아닌 금액 입력 시 예외 처리 추가.

test: 로또 구매 기능에 대한 테스트 코드 작성.

yebink-woowahan committed:
feat : 로또 금액을 담는 Money 클래스 생성

hojinDev committed:
feat(lotto): 로또 생성 및 구매 후 번호 출력 기능 추가

로또 생성 클래스와 구매 후 번호를 출력하는 기능을 추가. 로또 구매 시 로또 객체를 반환하도록 변경하고, 테스트 코드 수정 및 추가.

footer: JVM Toolchain 17로 버전 변경.

yebink-woowahan committed:
feat : 지난주 당첨 금액 입력 코드 작성

hojinDev committed:
feat(lotto): 당첨 결과 통계 및 수익률 계산 기능 추가

당첨 번호와 구매한 로또를 비교해 당첨 결과 통계 및 수익률을 계산하는 기능 추가.
Rank enum 생성으로 당첨 등수 및 상금을 관리하고 LottoResult 클래스에서 결과를 처리하도록 구현.

footer: 테스트 코드 및 README.md 업데이트.

yebink-woowahan committed:
refactor : 테스트 리팩토링 (BehaviorSpec -> FreeSpec)

yebink-woowahan committed:
refactor : 함수 라인 10줄 이내로 맞추기, 매직넘버 상수화

men7627 commented on src/main/kotlin/lotto/Lotto.kt:
```diff
@@ -0,0 +1,19 @@
+
+private const val MINIMUM_LOTTO_NUMBER = 1
+private const val MAXIMUM_LOTTO_NUMBER = 45
+private const val LOTTO_NUMBER_COUNT = 6
```
파일 상단에 상수를 선언하는 것은 전역적으로 사용할 때 그렇게 사용하는 것으로 알고 있는데 (어느 클래스에도 속하지 않기 때문에)
companion object로 클래스 관련 상수라는 의미를 사용하는게 좋아보이는데,
요렇게 하신 이유가 궁금합니다.

hjlee83 replied:
앗 이유는 깔끔해보여서 이렇게도 써보았는데요. 로또클래스에 관련 상수라서 companion object 로 변경하겠습니다.

men7627 commented on src/main/kotlin/lotto/Lotto.kt:
```diff
@@ -0,0 +1,19 @@
+class Lotto(input: List<Int>) {
+    val numbers = input
+
+    init {
+        input.forEach {
+            if (!(MINIMUM_LOTTO_NUMBER..MAXIMUM_LOTTO_NUMBER).contains(it)) {
+                throw IllegalArgumentException("1부터 45까지의 숫자를 입력하세요")
+            }
+        }
+    }
```
init 구문에서 유효성 검증을 함으로써, 생성자가 추가적으로 생겨도 항상 검증을 하게 한 포인트가 좋은 것 같습니다! 👍

hjlee83 replied:
감사합니다.

men7627 commented on src/main/kotlin/lotto/LottoShop.kt:
```diff
@@ -0,0 +1,19 @@
+        return lottoCount.downTo(1).map {
+            Lotto()
+        }
+    }
```
개인적인 의견으로는 객체지향적인 관점에서 보았을 때, 조금 어색한 네이밍 같습니다.

LottoShop은 Lotto를 생성하는 주체인것 같은데요
의미상 LottoShop.buyLotto는 로또샵이 로또를 사는것 같다는 느낌이들어서
Factory같은 느낌으로 Lotto를 생성해내는 네이밍으로 하는게 어떨까 싶습니다!

hjlee83 replied:
앗 넵 동의합니다. 로또를 판다는 의미로 sellLotto 요렇게 하려고 하는데요. 괜찮으실까요?

men7627 commented on src/main/kotlin/lotto/Money.kt:
```diff
@@ -0,0 +1,15 @@
+            throw IllegalArgumentException("올바른 금액을 입력하세요")
+        }
+        this.price = price.toInt()
+    }
```
Lotto 에서 작성해주신 것처럼 init 구문에서 유효성 검증을 하는게 더 좋지 않을까 합니다!

hjlee83 replied:
string 값을 넣어야해서. 값 검증은 init 변환검증은 invoke 로 넣을게요.

hjlee83 replied:
아 invoke 말고 보조생성자와 주생성자로 성공했습니다.

men7627 commented on src/main/kotlin/lotto/LottoGame.kt:
```diff
@@ -0,0 +1,38 @@
+    showLottoResult(lottos)
+}
+
+fun buyLotto(): List<Lotto> {
```
출력하는 기능들은 View 클래스로 구분하면 좋을 것 같습니다.

hjlee83 replied:
요거는 2차때 구현하겠습니다.

hojinDev committed:
refactor(lotto): 상수 위치를 companion object로 이동

로또 클래스의 가독성과 책임 분리를 위해 상수들을 companion object로 이동.
이를 통해 상수가 클래스 인스턴스와 무관한 공용 정보임을 명확히 표현.

footer: 코드 정리 및 클래스 구조 개선.

hojinDev committed:
fix(lotto): 예외 메시지에 상수 사용하도록 수정

예외 메시지에 상수를 활용하여 코드 중복을 제거하고 유지 보수성을 개선.
상수 값이 변경되더라도 메시지가 자동으로 반영되도록 처리.

footer: 예외 처리 가독성 향상.

hojinDev committed:
refactor(lotto): 금액 검증 로직 개선

금액 검증 로직을 init 및 invoke로 분리하여 가독성과 재사용성을 향상.
에러 메시지 및 최소 금액 상수를 추가하여 코드 중복 제거 및 유지 보수성 개선.

footer: 로직 재구성으로 유효성 검사 효율성 증가.

hojinDev committed:
refactor(lotto): 금액 파싱 및 검증 로직 개선

문자열 금액 입력 시 초기화 로직을 개선하고 중복된 검증 코드를 parseAndValidate 함수로 통합.
init 블록과 생성자에서 최소 금액 상수를 활용하여 코드 간소화.

footer: 코드 재사용성 및 유지 보수성 향상.

hojinDev committed:
refactor(lotto): buyLotto 메서드명을 sellLotto로 변경

메서드의 동작과 의도를 더 명확히 표현하기 위해 buyLotto를 sellLotto로 리팩터링.
이에 따라 관련 호출부와 테스트 코드도 동일하게 수정.

footer: 가독성 및 코드 일관성 강화.

hjlee83 commented:
아앗.

men7627 commented on src/test/kotlin/lotto/WinLottoTest.kt:
```diff
@@ -0,0 +1,29 @@
+
+class WinLottoTest : FreeSpec({
+
+    "문자가 포함된 경우 예외 발생" {
+        val exception =
+            shouldThrow<IllegalArgumentException> { WinLotto("1 2 3 4 5 육") }
+        exception.message shouldBe "올바른 숫자를 입력하세요"
+    }
+    "1보다 작거나 45보다 큰 숫자가 포함된 경우 예외 발생" {
+        val exception =
+            shouldThrow<IllegalArgumentException> { WinLotto("0 2 3 4 5 46") }
+        exception.message shouldBe "1부터 45까지의 숫자를 입력하세요"
+    }
+    "5개의 숫자를 입력하는 경우 예외 발생" {
+        val exception =
+            shouldThrow<IllegalArgumentException> { WinLotto("2 4 6 8 10") }
+        exception.message shouldBe "6개의 서로 다른 숫자를 입력하세요"
+    }
+    "6개 숫자 간 중복이 있는 경우" {
+        val exception =
+            shouldThrow<IllegalArgumentException> { WinLotto("2 4 6 8 10 2") }
+        exception.message shouldBe "6개의 서로 다른 숫자를 입력하세요"
+    }
```
💯
